stan/math/prim/fun/interp1.hpp

```diff
@@ -1,45 +1,48 @@
 #ifndef STAN_MATH_PRIM_FUN_INTERP1_HPP
 #define STAN_MATH_PRIM_FUN_INTERP1_HPP
 
-#include <stan/math/prim/meta.hpp>
+//#include <stan/math/prim/fun/sort_asc.hpp>
+
 using namespace std;
 namespace stan {
     namespace math {
-    /**
-     * Return the one-dimensional interpolation of the second argument of points corresponding to the first argument.
-     * @tparam T_true type of the true argument
-     * @tparam T_false type of the false argument
-     * @param c Boolean condition value.
-     * @param y_true Value to return if condition is true.
-     * @param y_false Value to return if condition is false.
-     */
-    inline Eigen::VectorXd
-    interp1(const Eigen::VectorXd& xData, const Eigen::VectorXd& yData, const Eigen::VectorXd& xVector)
-    {
-
-        Eigen::VectorXd yVals;
-        int size_x = xData.size();
-        int size_xData = xData.size();
+        template <typename T>
```

Comment:
I see you're already templating the type! That is good :D.

Start with only templating xTest though. Assume `xData` and `yData` are `doubles`.

Whenever you do calculations with multiple types, the output type is a function of all the different input types. I think Bob kinda talks about this like a least upper bound or something?

Like, if you have integers and doubles, and you do math, in general the output will be a double to not lose precision (cuz 1.5 isn't an integer, but 3 / 2.0 is 1.5).

Similarly if you have a float and a double and do math, in general the output should be a double again to not lose precision (the double is higher precision).

So in Stan if you do Math on a double and an autodiff type (which is why we template stuff), the output will be that autodiff type.

So we use the `return_type_t` metaprogram in Stan (this is specific for Stan) to compute return types.

So if you had the fully templated function we write that like this:

```cpp
template<typename T1, typename T2, typename T3>
inline std::vector<return_type_t<T1, T2, T3>>
interp1(const std::vector<T1>& xDataPre,
  const std::vector<T2>& yDataPre,
  const std::vector<T3>& xTest) {
  ...
}
```

But for now just start with only `xTest` templated:
```
template<typename T>
inline std::vector<T>
interp1(const std::vector<double>& xDataPre,
  const std::vector<double>& yDataPre,
  const std::vector<T>& xTest) {
  ...
}
```

This means all the temporaries you compute in Stan that are a function of a `T` should themselves have the `T` type.

Reply:
Thanks for the explanation; have changed xTest only <T> in the new version. Could I ask what could go wrong with the first(original) one please?

```
template<typename T>
inline std::vector<double> interp1(const std::vector<T>& xDataPre,
  const std::vector<T>& yDataPre,
  const std::vector<T>& xTest) {
  ...
}
```

```
template<typename T1, typename T2, typename T3>
inline std::vector<return_type_t<T1, T2, T3>>
interp1(const std::vector<T1>& xDataPre,
  const std::vector<T2>& yDataPre,
  const std::vector<T3>& xTest) {
  ...
}
```

Reply:
If all the template types are the same, then all three inputs need to be autodiff types.

In Stan, the functions need to be a mix of doubles or autodiff types. In this case there are 3 parameters, so 3 things that can either be doubles or autodiff types -> 8 total possible functions.